File: Day7.java
import java.util.*;
import java.io.*;
public class Day7 {
    static class Node {
        String name;
        int weight;
        int sum_of_child;
        List<String> children;
        List<Integer> child_weights;

        public Node() {
            sum_of_child = 0;
            children = new ArrayList<>();
            child_weights = new ArrayList<>();
        }
    }

    public static void main(String[] args) {
        String root = "gmcrj";
        Scanner s = new Scanner(System.in);
        HashMap<String, Node> map = new HashMap<>();
        while (s.hasNextLine()) {
            Node node = new Node();
            String line = s.nextLine();
            String[] tokens = line.split("->");
            String[] first = tokens[0].split("\\(");
            node.name = first[0].trim();
            node.weight = Integer.parseInt(first[1].split("\\)")[0]);
            if (tokens.length > 1) {
                String[] children = tokens[1].split("\\,");
                for (int i = 0; i < children.length; i ++)
                    node.children.add(children[i].trim());
            }
            map.put(node.name.trim(), node);
        }

        populate(map, root, 1);
    }

    public static void populate(HashMap<String, Node> map, String name, int depth) {
        Node node;
        if (name == null || (node = map.get(name.trim())) == null) return;
        int sum = 0;
        for (String c : node.children) {
            if (c == null) continue;
            Node child = map.get(c.trim());
            if (child == null) continue;
            populate(map, c.trim(), depth + 1);
            node.child_weights.add(child.weight);
            sum += child.weight;
        }
        if (!node.child_weights.isEmpty() && isDiff(node.child_weights)) {
            System.out.printf("%d %s %s %d\n", node.weight, name.trim(), listToString(node.child_weights), depth);
        }
        node.sum_of_child = sum;
        node.weight += sum;
        return;
    }

    public static boolean isDiff(List<Integer> list) {
        if (list.isEmpty())
            return false;
        int elem = list.get(0);
        for (int value : list) {
            if (elem != value)
                return true;
        }
        return false;
    }

    public static String listToString(List<Integer> list) {
        StringBuilder b = new StringBuilder();
        Iterator<Integer> iter = list.iterator();
        while(iter.hasNext()) {
            b.append(iter.next());
            if (iter.hasNext())
                b.append(",");
        }
        return b.toString();
    }
}

File: day7.js
var _ = require('underscore');
var fs = require('fs');
var filename = 'day7.txt';
fs.readFile(filename, 'utf8', function(err, data) {
  if (err) throw err;
  part1(data);
  part2(data);
});

function part1(data) {
    var arr = _.map(data.trim().split('\n'), (a) => {
        var pair = _.map(a.split('->'), (b) => {
            return b.trim();
        });
        var node = {};
        node.name = pair[0].split('(')[0].trim();
        if (pair.length > 1) {
            var children = _.map(pair[1].split(','), (c) => c.trim());
            node.children = {};
            _.each(children, (c) => {
                node.children[c] = true;
            })
        }
        return node;
    });
    var seen = {};
    _.each(arr, (node) => {
        _.each(node.children, (_, k) => {
            seen[k] = true;
        });
    });

    var unseen = [];
    _.each(arr, (node) => {
        if (!seen[node.name]) {
           unseen.push(node.name);
        }
    });
    console.log(unseen);
}

function part2(data) {
    var arr = _.map(data.trim().split('\n'), (a) => {
        var pair = _.map(a.split('->'), (b) => {
            return b.trim();
        });
        var node = {};
        var node_data = pair[0].split('(');
        node.name = node_data[0].trim();
        node.weight = node_data[1].split(')')[0];
        if (pair.length > 1) {
            var children = _.map(pair[1].split(','), (c) => c.trim());
            node.children = {};
            _.each(children, (c) => {
                node.children[c] = true;
            })
        }
        return node;
    });
    var tree = _.indexBy(arr, 'name');
    var root_node = 'gmcrj';
    var total_weights = {};
    fill_weight(tree, total_weights, root_node);
    find_weight(tree, 0, total_weights, root_node);
}

function fill_weight(tree, total_weights, root) {
    if (!tree || !root || !tree[root]) 
        return;
    var children = tree[root].children;
    total_weights[root] = tree[root].weight;
    _.each(children, (_, c) => {
        fill_weight(tree, total_weights, c);
        total_weights[root] += tree[c].weight;
    });
}

function find_weight(tree, current_weight, total_weights, root) {
    if (!tree[root])
        return;

    var child_weights = [];
    _.each(tree[root].children, (_, c) => {
        child_weights.push(tree[c].weight);
    });

    if (child_weights.length == 0)
        return;

    var same = child_weights[0];
    for (var elem of child_weights) {
        if (same != elem) {
            // uneven
            _.each(tree[root].children, (_, c) => {
                find_weight(tree, current_weight, total_weights, c);
            });
            break;
        }
    }
}
